### ISSUE TITLE
Add `chrome.*` APIs to support WebExtensions on chrome and FF.

### ISSUE THREAD

**Issue by qazbnm456** [qazbnm456/lulumi-browser#19]
We now have https://github.com/electron/electron/issues/5842 for Accessibility DevTools extension.
How about implementing more APIs to support WebExtensions as well to run those extensions that are on chrome and FF.

**Comment by qazbnm456** [comment]
**Currently Implemented APIs**

- [x] browserAction.onClicked
- [x] pageAction.show(tabId)
- [x] pageAction.hide(tabId)
- [x] pageAction.onClicked
- [x] alarms.get(name, callback)
- [x] alarms.getAll(callback)
- [x] alarms.clear(name, callback)
- [x] alarms.clearAll(callback)
- [x] alarms.create(name, alarmInfo)
- [x] alarms.onAlarm
- [x] commands.onCommand
- [x] runtime.id
- [x] runtime.getURL(path)
- [x] runtime.sendMessage(extensionId, message, options, responseCallback)
- [x] runtime.onMessage
- [x] extension.getURL(path)
- [x] extension.getBackgroundPage()
- [x] tabs.get(tabId, callback)
- [x] tabs.getCurrent(callback)
- [x] tabs.duplicate(tabId, callback)
- [x] tabs.query(queryInfo, callback)
- [x] tabs.update(tabId, updateProperties, callback)
- [x] tabs.reload(tabId, reloadProperties, callback)
- [x] tabs.remove(tabIds, callback)
- [x] tabs.detectLanguage(tabId, callback)
- [x] tabs.executeScript(tabId, details, callback)
- [x] tabs.insertCSS: (tabId, details, callback)
- [x] tabs.sendMessage(tabId, message, responseCallback)
- [x] tabs.onUpdated
- [x] tabs.onCreated
- [x] tabs.onRemoved
- [x] storage.local
- [x] storage.sync
- [x] storage.set(items, callback)
- [x] storage.get(keys, callback)
- [x] storage.onChanged
- [x] contextMenus.create(createProperties, callback)
- [x] webRequest.onBeforeRequest(callback, filter, opt_extraInfoSpec)
- [x] webRequest.onBeforeSendHeaders(callback, filter, opt_extraInfoSpec)
- [x] webRequest.onSendHeaders(callback, filter, opt_extraInfoSpec)
- [x] webRequest.onHeadersReceived(callback, filter, opt_extraInfoSpec)
- [x] webRequest.onResponseStarted(callback, filter, opt_extraInfoSpec)
- [x] webRequest.onBeforeRedirect(callback, filter, opt_extraInfoSpec)
- [x] webRequest.onCompleted(callback, filter, opt_extraInfoSpec)
- [x] webRequest.onErrorOccurred(callback, filter, opt_extraInfoSpec)
- [x] webNavigation.getFrame(details, callback)
- [x] webNavigation.getAllFrames(details, callback)
- [x] webNavigation.onBeforeNavigate(callback)
- [x] webNavigation.onCommitted(callback)
- [x] webNavigation.onDOMContentLoaded(callback)
- [x] webNavigation.onCompleted(callback)
- [x] webNavigation.onCreatedNavigationTarget(callback)

**Comment by qazbnm456** [comment]
**APIs needed for [uBlock](https://github.com/gorhill/uBlock)**

- [ ] i18n.getMessage
- [ ] runtime.getManifest
- [ ] localStorage.getItem

**Comment by qazbnm456** [comment]
**APIs needed for [thegreatsuspender](https://github.com/deanoemcke/thegreatsuspender)**

- [ ] contextMenu.removeAll

**Comment by qazbnm456** [comment]
**APIs needed for [Grammarly for Chrome](https://chrome.google.com/webstore/detail/grammarly-for-chrome/kbfnbcaeplbcioakkpcpgfkobkghlhen)**

- [ ] windows.onFocusChanged
- [ ] cookies.onChanged
- [ ] extension.isAllowedIncognitoAccess(callback)

Yout can extract the source code from .crx file through http://crxextractor.com/.

**Comment by jeromevalentin** [comment]
**APIs needed for [Virtual Keyboard](https://chrome.google.com/webstore/detail/virtualkeyboard/pflmllfnnabikmfkkaddkoolinlfninn)**

- [ ] extension.onRequest.addListener()
- [ ] extension.sendRequest()
- [X] extension.getURL()

**Comment by 82318179** [comment]
这个浏览器现在支持windows吗？

**Comment by qazbnm456** [comment]
@82318179 支持，但還有很多功能未完善。

**Comment by qazbnm456** [comment]
**APIs needed for [Steward](https://github.com/solobat/Steward)**

- [ ] extension.onRequest
- [x] extension.getURL(string path)
- [ ] extension.sendRequest(string extensionId, any request, function responseCallback)
- [ ] i18n.getMessage(string messageName, any substitutions)
- [ ] runtime.sendMessage(string extensionId, any message, object options, function responseCallback)
- [x] runtime.getManifest()
- [ ] storage.onChanged
- [ ] management.getAll(function callback)
- [ ] tabs.create(object createProperties, function callback)
- [ ] i18n.getUILanguage()

**Comment by samuelmaddock** [comment]
Hey @qazbnm456, I just discovered your project and was wondering if you had any plans to contribute these API additions back into the main electron repo.

**Comment by qazbnm456** [comment]
Hi @samuelmaddock, there's no such a plan to contribute these API additions back to electron currently since several dirty hacks have been done in this repo. But I'm thinking about making these API as a package to provide a way injecting them to existing electron apps like https://github.com/koush/electron-chrome.

Maybe after finishing the package mentioned above, I'll try to clean up the implementations and pick some back to upstream.

Thanks for your question.

**Comment by Sential** [comment]
Hi @qazbnm456. Have you managed how to impement `extension.getBackgroundPage()` in C++?

**Comment by qazbnm456** [comment]
Hi @Sential, I was busy before, so I didn't have the spare time figuring out how to implement it. I'll tackle this problem in the near future by mimicking the implementation of chrome.

**Comment by Sential** [comment]
Have you got Discord? Mine is Sential#9122. I want to talk about extensions API with you.

**Comment by qazbnm456** [comment]
@Sential I don't have one. I'll add and chat with you on weekend, perhaps, if I have time. Thank you!

**Comment by qazbnm456** [comment]
Been spending several days looking into extension system of chromium and found out that some extension APIs need lots of effort to be implemented, e.g. `extension.getBackgroundPage()` and `extension.getViews()`.

Would revisit them again after electron migrates to OOPIF architecture.

**Comment by samuelmaddock** [comment]
@qazbnm456 Checkout Muon if you need a reference. They have a good amount of the extensions API implemented. https://github.com/brave/muon

**Comment by Sential** [comment]
@qazbnm456 Have you managed to get uBlock Origin's cosmetic filtering to work?

**Comment by qazbnm456** [comment]
@stewartlord No, still have some problems making it work. It's more complicated than I expected and I'll look into it again when I'm available sometime.

**Comment by wangfei1988** [comment]
thanks

**Comment by wangfei1988** [comment]
可以支持安卓 和ios吗

**Comment by qazbnm456** [comment]
@wangfei1988 很抱歉，沒有喔！Lulumi 底層是 Electron，而 Electron 沒有支援 Android 和 IOS。

**Comment by samuelmaddock** [comment]
@qazbnm456 i wanted to point out that i've been contributing to electron to add better support for chrome extensions https://github.com/electron/electron/pulls/samuelmaddock

soon i'll be looking into supporting additional `chrome.*` APIs. send me an email, listed on my profile, if you'd be interested in helping out. 🙂

**Comment by qazbnm456** [comment]
Hi @samuelmaddock, it's great to have you in Electron team so as to push the implementation of `chrome.*` APIs better.

**Comment by minpoo** [comment]
点击关闭浏览器后，进程没有被杀死，在window上，第二次打开导致失败，怎么解决
